---
title: "01 Analysis of BL patient dataset from Schmitz et al. Nature 2012 "
author: "Joanna Krupka"  
date: "October 16, 2019"
output: 
  html_document:
    theme: cosmo
    code_folding: show
---

## Introduction

We used RNA-Seq datasets from Schmitz et al., 2012 to examine gene expression pattern associated with DDX3X mutation

## Objectives

Perform differential expression and GSEA analysis of BL patients RNA-Seq data

## Material & Methods

```{r eval = T, echo = T, message = F, warning = F, cache = F, error = F}
library(tidyverse)
library(DESeq2)
library(biomaRt)
library(pheatmap)
library(msigdbr)
library(clusterProfiler)
library(ReactomePA)
library(enrichplot)
library(fgsea)
library(ggrepel)
library(org.Hs.eg.db)

NATURE2012COUNTS <- "../utilis/Schmitz2012/nature2012_counts_new.RData"
NATURE2012META <- "../utilis/Schmitz2012/BL_Nature2012_metadata.csv"

source("../utilis.R")

# Load data and metadata
load(NATURE2012COUNTS)
metadata <- read_csv(NATURE2012META) 
```

# Exploratory data analysis

```{r}
# Prepare data, remove smaples failing QC
patientsMeta <- metadata %>%
  mutate(readLength = as.character(readLength)) %>%
  filter(sample_type == "patient" & ID != "SRR387407") %>%
  dplyr::select(ID, condition, mut_details, instrument, LoadDate, readLength) 

patientsCounts <- counts$counts 
colnames(patientsCounts) <- splitvec(splitvec(colnames(patientsCounts), "[.]", 11), "_", 1)

# Annotations
if (!exists("biomart_ensembl2name")){
  ensembl <- useMart("ensembl", dataset="hsapiens_gene_ensembl", host = "http://jul2019.archive.ensembl.org")
  biomart_ensembl2name <- getBM(attributes=c('ensembl_gene_id', 
                                             'external_gene_name', 
                                             'entrezgene_id'), 
                                filters = "ensembl_gene_id", 
                                values = splitvec(rownames(patientsCounts), "[.]", 1), 
                                mart = ensembl)
}

patientsCounts <- patientsCounts %>% as.data.frame() %>%
  dplyr::select(one_of(patientsMeta$ID)) %>%
  rownames_to_column("gene_id") %>%
  na.omit() %>%
  remove_rownames() %>%
  column_to_rownames("gene_id")
```

## Analysis 

### Differential expression

```{r eval = T, echo = T, message = F, warning = F, cache = F, error = F}
patientsCountsFiltered <- remove_low(patientsCounts, 128, 0.25)
dds <- DESeqDataSetFromMatrix(countData = patientsCounts,
                              colData = patientsMeta,
                              design = ~ condition)

dds$condition <- relevel(dds$condition, ref = "wt")
dds <- DESeq(dds)
resultsNames(dds)
res <- results(dds)

resDf <- as.data.frame(res) %>%
  rownames_to_column("ensembl_gene_id") %>%
  mutate(ensembl_gene_id = splitvec(ensembl_gene_id, "[.]", 1)) %>%
  left_join(biomart_ensembl2name) %>%
  dplyr::select(ensembl_gene_id, external_gene_name, baseMean, log2FoldChange, 
                lfcSE, stat, pvalue, padj, entrezgene_id) %>%
  na.omit()

# VST normalize
dds <- DESeqDataSetFromMatrix(countData = patientsCountsFiltered,
                              colData = patientsMeta,
                              design = ~ condition)
dds$condition <- relevel(dds$condition, ref = "wt")
BLvst <- assay(vst(dds)) %>% as.data.frame() %>%
  rownames_to_column("ensembl_gene_id")

# Rlog normalise
BLrld <- assay(rlog(dds)) %>% as.data.frame() %>%
  rownames_to_column("ensembl_gene_id")

toSave <- resDf  %>%
  mutate(status = case_when(
    padj <= 0.1 & log2FoldChange > 0.4 ~ "mRNA_up",
    padj <= 0.11 & log2FoldChange < -0.4 ~ "mRNA_down",
    TRUE ~ "stable"
  ))

#write_csv(toSave, "results/BLpatients_DESeq2_sumtab.csv")
write_csv(BLvst, "../utilis/Schmitz2012/Schmitz2012_RNAseq_VST.csv")
write_csv(BLrld, "../utilis/Schmitz2012/Schmitz2012_RNAseq_RLD.csv")
```

# GSEA

# GSEA desktop input  

```{r eval = F, echo = T, message = F, warning = F, cache = F, error = F}
# BL patients
#ExpressionSet construction
norm_counts <- BLrld %>%
  as.data.frame() %>%
  gather(sample, VST, -ensembl_gene_id) %>%
  mutate(ensembl_gene_id = splitvec(ensembl_gene_id, "[.]", 1)) %>%
  left_join(biomart_ensembl2name) %>%
  group_by(sample, external_gene_name) %>%
  summarize(VST = mean(VST)) %>%
  filter(!duplicated(external_gene_name)) %>%
  na.omit() %>%
  ungroup() %>%
  spread(sample, VST) %>%
  remove_rownames() %>%
  column_to_rownames("external_gene_name")

#AnnotatedDataFrame with metadata
meta <- data.frame(labelDescription= colnames(patientsMeta[,1:2]),
                   row.names=colnames(patientsMeta[,1:2]))
pD <- patientsMeta[,1:2] %>% as.data.frame()
rownames(pD) <- patientsMeta$ID
phenoData <- new("AnnotatedDataFrame", data=pD, varMetadata=meta)

#AnnotatedDataFrame with features data
features <- biomart_ensembl2name[,c(1,2)][match(rownames(norm_counts), biomart_ensembl2name$external_gene_name),]
rownames(features) <- rownames(norm_counts)
meta <- data.frame(labelDescription= colnames(features),
                       row.names=colnames(features))
featureData <- new("AnnotatedDataFrame", data=features, varMetadata=meta)

#ExpressionSet object
exampleSet <- ExpressionSet(assayData = as.matrix(norm_counts), 
                            phenoData = phenoData, 
                            featureData = featureData)

#create GSEA files
ArrayTools::createGSEAFiles(mydir = paste(getwd(), "GSEA_Schmitz2012", sep="/"), exampleSet, "condition")
```


## BL patients  

```{r eval = T, echo = T, message = F, warning = F, cache = F, error = F}
set.seed(1)
mSigH <- msigdbr(species = "Homo sapiens", category = "H")  %>% 
  dplyr::select(gs_name, entrez_gene)

# Load pre-ranked genes
rankedList <- read_csv("GSEA_Schmitz2012/ranked_gene_list_mut_versus_wt.csv") %>%
  left_join(biomart_ensembl2name, by = c("NAME" = "external_gene_name")) %>%
  na.omit()

geneList <- rankedList$SCORE
names(geneList) <- as.character(rankedList$entrezgene_id)

BL_H <- GSEA(geneList, TERM2GENE = mSigH, 
             minGSSize = 8, pvalueCutoff = 0.25, seed = 1, exponent = 1)
BL_H <- setReadable(BL_H, org.Hs.eg.db, keyType = "ENTREZID")
BL_Htab <- BL_H %>% as.data.frame()

write_csv(BL_Htab, "GSEA_BL/BL_GSEA_Htab.csv")

IDs <- c("HALLMARK_MYC_TARGETS_V1",
         "HALLMARK_UNFOLDED_PROTEIN_RESPONSE",
         "HALLMARK_MTORC1_SIGNALING")
gseaplot2(BL_H, geneSetID = IDs, pvalue_table = TRUE,
          color = divergingPal[c(9,10,11)],
          rel_heights = c(1.5, 0.5, 0.5))

ggsave("plots/Schmitz2012_HALLMARK.pdf")
```

#### Added (): integrated GSEA analysis  

```{r eval = T, echo = T, message = F, warning = F, cache = F, error = F}
# Load signal to noise ratios computed by Broad Institute GSEA app
BL_vst <- read_delim("GSEA_Schmitz2012/Result/BLpatients.probe.gct", delim = "\t", skip = 2)
U2392clones_vst <- read_delim("../09_U2932clones/GSEA_clones/U2932clones.probe.gct", delim = "\t", skip = 2)
U2392sh_vst <- read_delim("../04_RiboSeq/GSEA_sh/Result/U2932/GSEA/Treatment2.probe.gct", delim = "\t", skip = 2)

BL_meta <- patientsMeta
U2392clone_meta <- read_csv("../utilis/expressionData/U2932clones/U2932clones_metadata.csv")
U2392sh_meta <- read_csv("../utilis/expressionData/DDX3Xsh/RiboRNA_DDX3Xsh.csv")

# Select genes to plot (unfolded protein response)  
coreGenes <- read_csv("UPR_coreEnrichment_allExp.csv")
coreGenes <- strsplit(coreGenes$UPR_coreEnrichment, split = "/")

# Bl biopsies
plot_df <- BL_vst %>%
  filter(NAME %in% coreGenes[[1]]) %>%
  dplyr::select(-NAME) %>%
  column_to_rownames("Description")

all(BL_meta$ID == colnames(plot_df))
## Annotations
annotation_col <- tibble(DXstatus = BL_meta$condition,
                         ID = BL_meta$ID,
                         MeanExpr = colMeans(plot_df)) %>%
  filter(ID %in% colnames(plot_df)) %>%
  arrange(DXstatus, MeanExpr) %>%
  column_to_rownames("ID") %>%
  dplyr::select(-MeanExpr)

Var1 = brewer.pal(5,"Dark2")[c(1,2)]
names(Var1) = unique(annotation_col$DXstatus) 

ann_colors = list(DXstatus = Var1) 

colors <- c(-3, seq(-2.99,2.99,by=0.01),3) 
col <- brewer.pal(5,"RdBu")[c(1,5)]
palette <- rev(colorRampPalette(c(col[1], "white", col[2]))(n = length(colors)))
order <- rownames(plot_df)[order(rowMeans(plot_df))]

ph1 <- pheatmap(plot_df[order,rownames(annotation_col)],
         cluster_cols = F, 
         cluster_rows = F,
         scale = "row", 
         color = palette, 
         breaks = colors,
         annotation_col = annotation_col, 
         annotation_colors = ann_colors,
         filename = "plots/Schmitz2012_UPRgenes.pdf", 
         cellwidth = 5,
         cellheight = 10,
         fontsize_row = 8,
         show_colnames = F,
         border_color = NA)

# U2392 clones
plot_df <- U2392clones_vst %>%
  filter(NAME %in% coreGenes[[2]]) %>%
  dplyr::select(-NAME) %>%
  column_to_rownames("Description")
U2392clone_meta <- U2392clone_meta %>%
  filter(Sample %in% colnames(plot_df))
all(U2392clone_meta$Sample == colnames(plot_df))

## Annotations
annotation_col <- tibble(DXstatus = U2392clone_meta$Condition,
                         ID = U2392clone_meta$Sample,
                         MeanExpr = colMeans(plot_df)) %>%
  filter(ID %in% colnames(plot_df)) %>%
  arrange(DXstatus, MeanExpr) %>%
  column_to_rownames("ID") %>%
  dplyr::select(-MeanExpr)

Var1 =  brewer.pal(5,"Dark2")[c(3,4)]
names(Var1) = unique(annotation_col$DXstatus) 

ann_colors = list(DXstatus = Var1) 

colors <- c(-2, seq(-1.99,1.99,by=0.01),2) 
col <- brewer.pal(5,"RdBu")[c(1,5)]
palette <- rev(colorRampPalette(c(col[1], "white", col[2]))(n = length(colors)))

ph2 <- pheatmap(plot_df[order(rowMeans(plot_df)),rownames(annotation_col)],
         cluster_cols = F, 
         cluster_rows = F,
         scale = "row", 
         color = palette, 
         breaks = colors,
         annotation_col = annotation_col, 
         annotation_colors = ann_colors,
         filename = "plots/U2392clones_UPRgenes.pdf",
         cellwidth = 5,
         cellheight = 10,
         fontsize_row = 8,
         show_colnames = F,
         border_color = NA)

# U2392 sh
plot_df <- U2392sh_vst %>%
  filter(NAME %in% coreGenes[[3]]) %>%
  dplyr::select(-NAME) %>%
  column_to_rownames("Description")

U2392sh_meta <- U2392sh_meta %>%
  filter(Sample %in% colnames(plot_df)) %>%
  arrange(match(Sample, colnames(plot_df)))
all(U2392sh_meta$Sample == colnames(plot_df))

## Annotations
annotation_col <- tibble(DXstatus = factor(U2392sh_meta$Treatment2, levels = c("sh", "ctl")),
                         ID = U2392sh_meta$Sample,
                         MeanExpr = colMeans(plot_df)) %>%
  filter(ID %in% colnames(plot_df)) %>%
  arrange(DXstatus,MeanExpr) %>%
  column_to_rownames("ID") %>%
  dplyr::select(-MeanExpr)

Var1 = brewer.pal(6,"Dark2")[c(5,6)]
names(Var1) = unique(annotation_col$DXstatus) 

ann_colors = list(DXstatus = Var1) 

colors <- c(-2, seq(-1.99,1.99,by=0.01),2) 
col <- brewer.pal(5,"RdBu")[c(1,5)]
palette <- rev(colorRampPalette(c(col[1], "white", col[2]))(n = length(colors)))

ph2 <- pheatmap(plot_df[order(rowMeans(plot_df)),rownames(annotation_col)],
         cluster_cols = F, 
         cluster_rows = F,
         scale = "row", 
         color = palette, 
         breaks = colors,
         annotation_col = annotation_col, 
         annotation_colors = ann_colors,
         filename = "plots/U2392sh_UPRgenes.pdf",
         cellwidth = 5,
         cellheight = 10,
         fontsize_row = 8,
         show_colnames = F,
         border_color = NA)
```

###  U2932 clones   

```{r eval = T, echo = T, message = F, warning = F, error = F}  
rankedList <- read_csv("../09_U2932clones/GSEA/U2932clones_signalToNoise_rank.csv") %>%
  left_join(biomart_ensembl2name, by = c("NAME" = "external_gene_name")) %>%
  na.omit()

U2932clonesSumtab <- read_csv("../09_U2932clones/tables/U2932clones_RNA_sumtab.csv")

geneList_cl <- rankedList$SCORE
names(geneList_cl) <- as.character(rankedList$entrezgene_id)

U2932clones_H <- GSEA(geneList_cl, TERM2GENE = mSigH, 
             minGSSize = 8, pvalueCutoff = 1, seed = 1, exponent = 1)
U2932clones_H <- setReadable(U2932clones_H, org.Hs.eg.db, keyType = "ENTREZID")
U2932clones_Htab <- U2932clones_H %>% as.data.frame()

write_csv(BL_Htab, "GSEA_Schmitz2012/U2932clones_GSEA_Htab.csv")

IDs <- c("HALLMARK_MYC_TARGETS_V1",
         "HALLMARK_UNFOLDED_PROTEIN_RESPONSE",
         "HALLMARK_MTORC1_SIGNALING")

gseaplot2(U2932clones_H, geneSetID = IDs, pvalue_table = TRUE,
          color = divergingPal[c(9,10,11)],
          rel_heights = c(1.5, 0.5, 0.5)) 
ggsave("plots/U2932clones_HALLMARK.pdf")
```
